File: tests/java/org/elastos/hive/did/AppDID.java
package org.elastos.hive.did;

import org.elastos.did.VerifiableCredential;
import org.elastos.did.VerifiablePresentation;
import org.elastos.did.exception.DIDException;
import org.elastos.did.jwt.Header;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;
import java.util.List;

public class AppDID extends DIDEntity {

	private String appId = "appId";

	public AppDID(String name, String mnemonic, String phrasepass, String storepass) throws DIDException {
		super(name, mnemonic, phrasepass, storepass);
	}

	public String getAppDid() {
		return appId;
	}

	public VerifiablePresentation createPresentation(VerifiableCredential vc, String realm, String nonce) throws DIDException {
		VerifiablePresentation.Builder vpb = VerifiablePresentation.createFor(getDid(), getDIDStore());
		List<VerifiableCredential> vcs = new ArrayList<VerifiableCredential>(1);
		vcs.add(vc);

		VerifiablePresentation vp = vpb.credentials(vcs.toArray(new VerifiableCredential[vcs.size()]))
				.realm(realm)
				.nonce(nonce)
				.seal(getStorePassword());

		System.out.println("VerifiablePresentation:");
		String vpStr = vp.toString();
		System.out.println(vpStr);

		return vp;
	}

	public String createToken(VerifiablePresentation vp, String hiveDid) throws DIDException {

		Calendar cal = Calendar.getInstance();
		cal.set(Calendar.MILLISECOND, 0);
		Date iat = cal.getTime();
		Date nbf = cal.getTime();
		cal.add(Calendar.MONTH, 3);
		Date exp = cal.getTime();

		// Create JWT token with presentation.
		String token = getDocument().jwtBuilder()
				.addHeader(Header.TYPE, Header.JWT_TYPE)
				.addHeader("version", "1.0")

				.setSubject("DIDAuthResponse")
				.setAudience(hiveDid)
				.setIssuedAt(iat)
				.setExpiration(exp)
				.setNotBefore(nbf)
				.claimWithJson("presentation", vp.toString())
				.sign(storepass)
				.compact();

		System.out.println("JWT Token:");
		System.out.println("  " + token);
		return token;
	}
}

File: tests/src/did/appdid.ts
import { VerifiableCredential, VerifiablePresentation, DIDDocument, JWTHeader, DefaultDIDAdapter, DIDBackend, VerificationEventListener } from "@elastosfoundation/did-js-sdk";
import dayjs from "dayjs";
import { DIDEntity } from "./didentity";

export class AppDID extends DIDEntity {
	private appId = "appId";

	public constructor(name: string, mnemonic: string, phrasepass: string, storepass: string, did: string) {
		super(name, mnemonic, phrasepass, storepass, did);
	}

	public static async create(name: string, mnemonic: string, phrasepass: string, storepass: string, did?: string): Promise<AppDID> {
		DIDBackend.initialize(new DefaultDIDAdapter("mainnet")); 
		//DIDBackend.initialize(new DefaultDIDAdapter("https://api-testnet.elastos.io/newid")); 
        let newInstance = new AppDID(name, mnemonic, phrasepass, storepass, did);
		await newInstance.initPrivateIdentity(mnemonic);	
		await newInstance.initDid();

        return newInstance;
    }

	public getAppDid(): string {
		return this.getDid().toString();
	}

	public async createPresentation(vc: VerifiableCredential, realm: string, nonce: string): Promise<VerifiablePresentation> {

		
		let vpb = await VerifiablePresentation.createFor(this.getDid(), null, this.getDIDStore());
		let vp = await vpb.credentials(vc)
				.realm(realm)
				.nonce(nonce)
				.seal(this.getStorePassword());

		
			
				
		AppDID.LOG.info("VerifiablePresentation:{}", vp.toString());

		let listener = VerificationEventListener.getDefaultWithIdent("isValid");
		AppDID.LOG.trace("VerifiablePresentation is Valid :{}", await vp.isValid(listener));
		AppDID.LOG.trace("Listener :{}", listener.toString());

		return vp;
	}

	public async createToken(vp: VerifiablePresentation, hiveDid: string): Promise<string> {
        let cal = dayjs();
        let iat = cal.unix();
        let nbf = cal.unix();
        let exp = cal.add(3, 'month').unix();

		// Create JWT token with presentation.
		let doc: DIDDocument = await this.getDocument();
		let token = await doc.jwtBuilder()
				.addHeader(JWTHeader.TYPE, JWTHeader.JWT_TYPE)
				.addHeader("version", "1.0")
				.setSubject("DIDAuthResponse")
				.setAudience(hiveDid)
				.setIssuedAt(iat)
				.setExpiration(exp)
				.setNotBefore(nbf)
				.claimsWithJson("presentation", vp.toString(true))
				.sign(this.storepass);

		AppDID.LOG.info("JWT Token: {}", token);
		return token;
	}
}
